### PR TITLE
[SPARK-25502][CORE][WEBUI]Empty Page when page number exceeds the reatinedTask size.

### PR DESCRIPTION
**Pull request by shahidki31**
## What changes were proposed in this pull request?
Test steps :    
1)  bin/spark-shell --conf spark.ui.retainedTasks=200
```
val rdd = sc.parallelize(1 to 1000, 1000)
rdd.count
```

Stage tab in the UI will display 10 pages with 100 tasks per page. But number of retained tasks is only 200. So, from the 3rd page onwards will display nothing. 
 We have to calculate total pages based on the number of tasks need display in the UI. 

**Before fix:**
empty_4

**After fix:**
empty_3

## How was this patch tested?

Manually tested

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @vanzin

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 1]
```diff
@@ -685,7 +685,15 @@ private[ui] class TaskDataSource(
 
   private var _tasksToShow: Seq[TaskData] = null
 
-  override def dataSize: Int = taskCount(stage)
+  override def dataSize: Int = {
+    val storedTasks = store.taskCount(stage.stageId, stage.attemptId).toInt
+    val totalTasks = taskCount(stage)
+    if (totalTasks > storedTasks) {
```
[the remark this answers is not in the document]
Comment (shahidki31) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Yes. totalTasks will be always greater than or equal to  storedTasks. We can simply return storedTasks. But for better understanding I have put it in the if else condition.

I have modified the code based on your suggestion.

[the remark this answers is not in the document]
Comment (srowen) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Oh, hm I would have thought it's simpler to return the single value that this always takes on.

**Review by srowen — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (shahidki31) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Thanks. I have updated the PR  by  returning only storedTasks. Kindly look into the updated code.

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 2]
```diff
@@ -685,7 +685,10 @@ private[ui] class TaskDataSource(
 
   private var _tasksToShow: Seq[TaskData] = null
 
-  override def dataSize: Int = taskCount(stage)
+  override def dataSize: Int = {
+    val storedTasks = store.taskCount(stage.stageId, stage.attemptId).toInt
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: You don't need the intermediate variable here.

Comment (shahidki31) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Done. Thanks

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 3]
```diff
@@ -685,7 +685,7 @@ private[ui] class TaskDataSource(
 
   private var _tasksToShow: Seq[TaskData] = null
 
-  override def dataSize: Int = taskCount(stage)
+  override def dataSize: Int = store.taskCount(stage.stageId, stage.attemptId).toInt
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: nit: after this change, the function `taskCount()` is only referenced by `totalTasks`, we can inline that.

Comment (shahidki31) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Thanks. Done.

**Review by srowen — APPROVED** [review]

**Comment by vanzin** [comment]
Merging to master / 2.4 / 2.3.

**Pull request closed without merging**